exonum/src/runtime/dispatcher/types.rs

```diff
@@ -0,0 +1,164 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Internal dispatcher data types
+
+use exonum_proto::ProtobufConvert;
+
+use std::fmt;
+
+use crate::{
+    proto::schema,
+    runtime::{ArtifactSpec, InstanceSpec},
+};
+
+/// Status of an artifact deployment.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum ArtifactStatus {
+    /// The artifact is pending deployment.
+    Pending = 0,
+    /// The artifact has been successfully deployed.
+    Active = 1,
+}
+
+impl fmt::Display for ArtifactStatus {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Active => f.write_str("active"),
+            Self::Pending => f.write_str("pending"),
+        }
+    }
+}
+
+impl ProtobufConvert for ArtifactStatus {
+    type ProtoStruct = schema::dispatcher::ArtifactStatus;
+
+    fn to_pb(&self) -> Self::ProtoStruct {
+        match self {
+            Self::Active => Self::ProtoStruct::ARTIFACT_ACTIVE,
+            Self::Pending => Self::ProtoStruct::ARTIFACT_PENDING,
+        }
+    }
+
+    fn from_pb(pb: Self::ProtoStruct) -> Result<Self, failure::Error> {
+        Ok(match pb {
+            Self::ProtoStruct::ARTIFACT_ACTIVE => Self::Active,
+            Self::ProtoStruct::ARTIFACT_PENDING => Self::Pending,
+        })
+    }
+}
+
+/// Status of a service instance.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum InstanceStatus {
+    /// The service instance is pending deployment.
+    Pending = 0,
+    /// The service instance has been successfully deployed.
+    Active = 1,
+}
+
+impl fmt::Display for InstanceStatus {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Active => f.write_str("active"),
+            Self::Pending => f.write_str("pending"),
+        }
+    }
+}
+
+impl ProtobufConvert for InstanceStatus {
+    type ProtoStruct = schema::dispatcher::InstanceStatus;
+
+    fn to_pb(&self) -> Self::ProtoStruct {
+        match self {
+            Self::Active => Self::ProtoStruct::SERVICE_ACTIVE,
+            Self::Pending => Self::ProtoStruct::SERVICE_PENDING,
+        }
+    }
+
+    fn from_pb(pb: Self::ProtoStruct) -> Result<Self, failure::Error> {
+        Ok(match pb {
+            Self::ProtoStruct::SERVICE_ACTIVE => Self::Active,
+            Self::ProtoStruct::SERVICE_PENDING => Self::Pending,
+        })
+    }
+}
+
+/// Current state of artifact in dispatcher.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, ProtobufConvert, BinaryValue, ObjectHash)]
+#[protobuf_convert(source = "schema::dispatcher::ArtifactState")]
+pub struct ArtifactState {
+    /// Artifact specification.
+    pub spec: ArtifactSpec,
+    /// Artifact deployment status.
+    pub status: ArtifactStatus,
+}
+
+impl ArtifactState {
+    /// Returns underlining artifact spec if status is active.
+    pub fn active(self) -> Option<ArtifactSpec> {
```

Comment:
Nit: IMO, `active` / `pending` are too vague to be public. I think if they were renamed to something `spec_if_active` / `spec_if_pending`, their purpose would become clearer. Alternatively, might make sense to have a method like

```rust
fn filter_by_status(self, status: ArtifactStatus) -> Option<ArtifactSpec> {}
```